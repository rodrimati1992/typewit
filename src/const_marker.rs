//! Marker types for passing constants as type arguments.
//! 
//! # Example
//! 
//! This example emulates specialization,
//! eliding a `.clone()` call when the created array is only one element long.
//! 
//! ```rust
//! use typewit::{const_marker::Usize, TypeCmp, TypeEq};
//! 
//! let arr = [3u8, 5, 8];
//! 
//! assert_eq!(repeat(3), []);
//! assert_eq!(repeat(3), [3]);
//! assert_eq!(repeat(3), [3, 3]);
//! assert_eq!(repeat(3), [3, 3, 3]);
//! assert_eq!(repeat(3), [3, 3, 3, 3]);
//! 
//! 
//! fn repeat<T: Clone, const OUT: usize>(val: T) -> [T; OUT] {
//!     // `te_len` Ã¬s a `TypeEq<Usize<OUT>, Usize<1>>`
//!     if let TypeCmp::Eq(te_len) = Usize::<OUT>.equals(Usize::<1>) {
//!         // This branch is ran when `OUT == 1`
//!         TypeEq::new::<T>()    // returns `TypeEq<T, T>`
//!             .in_array(te_len) // returns `TypeEq<[T; OUT], [T; 1]>`
//!             .to_left([val])   // goes from `[T; 1]` to `[T; OUT]`
//!     } else {
//!         // This branch is ran when `OUT != 1`
//!         [(); OUT].map(|_| val.clone())
//!     }
//! }
//! ```
//! 
//! 

use crate::{
    TypeEq,
    TypeNe,
};

mod const_witnesses;

pub use const_witnesses::*;

#[cfg(feature = "adt_const_marker")]
mod slice_const_markers;

#[cfg(feature = "adt_const_marker")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "adt_const_marker")))]
pub use slice_const_markers::Str;

/// Marker types for `const FOO: &'static [T]` parameters.
#[cfg(feature = "adt_const_marker")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "adt_const_marker")))]
pub mod slice {
    pub use super::slice_const_markers::{
        BoolSlice,
        CharSlice,
        U8Slice,
        U16Slice,
        U32Slice,
        U64Slice,
        U128Slice,
        UsizeSlice,
        I8Slice,
        I16Slice,
        I32Slice,
        I64Slice,
        I128Slice,
        IsizeSlice,
        StrSlice,
    };
}

struct Helper<L, R>(L, R);



macro_rules! __const_eq_with {
    ($L:ident, $R:ident) => {
        $L == $R
    };
    ($L:ident, $R:ident, ($L2:ident, $R2:ident) $cmp:expr) => ({
        let $L2 = $L;
        let $R2 = $R;
        $cmp
    });
} pub(crate) use __const_eq_with;

macro_rules! declare_const_param_type {
    (
        $(#[$struct_docs:meta])*
        $struct:ident($prim:ty)

        $(
            $(#[$eq_docs:meta])*
            fn equals $(($L:ident, $R:ident) $comparator:block)?;
        )?
    ) => {
        #[doc = concat!(
            "Marker type for passing `const VAL: ", stringify!($prim),
            "` as a type parameter."
        )]
        $(#[$struct_docs])*
        #[derive(Debug, Copy, Clone)]
        pub struct $struct<const VAL: $prim>;

        impl<const L: $prim, const R: $prim> $crate::const_marker::Helper<$struct<L>, $struct<R>> {
            const EQ: Result<
                TypeEq<$struct<L>, $struct<R>>,
                TypeNe<$struct<L>, $struct<R>>,
            > = if crate::const_marker::__const_eq_with!(
                L,
                R
                $($(, ($L, $R) $comparator)?)?
            ) {
                // SAFETY: `L == R` (both are std types with sensible Eq impls)
                // therefore `$struct<L> == $struct<R>`
                unsafe {
                    Ok(TypeEq::<$struct<L>, $struct<R>>::new_unchecked())
                }
            } else {
                // SAFETY: `L != R` (both are std types with sensible Eq impls)
                // therefore `$struct<L> != $struct<R>`
                unsafe {
                    Err(TypeNe::<$struct<L>, $struct<R>>::new_unchecked())
                }
            };

            const EQUALS: crate::TypeCmp<$struct<L>, $struct<R>> = match Self::EQ {
                Ok(x) => crate::TypeCmp::Eq(x),
                Err(x) => crate::TypeCmp::Ne(x),
            };
        }

        impl<const VAL: $prim> $struct<VAL> {
            /// Compares `self` and `other` for equality.
            ///
            /// Returns:
            /// - `Ok(TypeEq)`: if `VAL == OTHER`
            /// - `Err(TypeNe)`: if `VAL != OTHER`
            ///
            #[inline(always)]
            #[deprecated(note = "superceeded by `equals` method", since = "1.8.0")]
            pub const fn eq<const OTHER: $prim>(
                self, 
                _other: $struct<OTHER>,
            ) -> Result<
                TypeEq<$struct<VAL>, $struct<OTHER>>,
                TypeNe<$struct<VAL>, $struct<OTHER>>,
            > {
                $crate::const_marker::Helper::<$struct<VAL>, $struct<OTHER>>::EQ
            }

            /// Compares `self` and `other` for equality.
            ///
            /// Returns:
            /// - `TypeCmp::Eq(TypeEq)`: if `VAL == OTHER`
            /// - `TypeCmp::Ne(TypeNe)`: if `VAL != OTHER`
            ///
            $($(#[$eq_docs])*)?
            #[inline(always)]
            pub const fn equals<const OTHER: $prim>(
                self, 
                _other: $struct<OTHER>,
            ) -> crate::TypeCmp<$struct<VAL>, $struct<OTHER>> {
                $crate::const_marker::Helper::<$struct<VAL>, $struct<OTHER>>::EQUALS
            }
        }
    };
} pub(crate) use declare_const_param_type;


declare_const_param_type!{
    Bool(bool)

    /// 
    /// For getting a type witness that
    /// `Bool<B>` is either `Bool<true>` or `Bool<false>`,
    /// you can use [`BoolWit`].


    /// 
    fn equals;
}
declare_const_param_type!{Char(char)}

declare_const_param_type!{U8(u8)}
declare_const_param_type!{U16(u16)}
declare_const_param_type!{U32(u32)}
declare_const_param_type!{U64(u64)}
declare_const_param_type!{U128(u128)}

declare_const_param_type!{
    Usize(usize)

    /// # Examples
    /// 
    /// ### Array
    /// 
    /// This example demonstrates how `Usize` can be used to 
    /// specialize behavior on array length.
    /// 
    /// (this example requires Rust 1.61.0, because it uses trait bounds in const fns)
    #[cfg_attr(not(feature = "rust_1_61"), doc = "```ignore")]
    #[cfg_attr(feature = "rust_1_61", doc = "```rust")]
    /// use typewit::{const_marker::Usize, TypeCmp, TypeEq};
    /// 
    /// assert_eq!(try_from_pair::<_, 0>((3, 5)), Ok([]));
    /// assert_eq!(try_from_pair::<_, 1>((3, 5)), Ok([3]));
    /// assert_eq!(try_from_pair::<_, 2>((3, 5)), Ok([3, 5]));
    /// assert_eq!(try_from_pair::<_, 3>((3, 5)), Err((3, 5)));
    /// 
    /// 
    /// const fn try_from_pair<T: Copy, const LEN: usize>(pair: (T, T)) -> Result<[T; LEN], (T, T)> {
    ///     if let TypeCmp::Eq(te_len) = Usize::<LEN>.equals(Usize::<0>) {
    ///         // this branch is ran on `LEN == 0`
    ///         // `te_len` is a `TypeEq<Usize<LEN>, Usize<0>>`
    ///         Ok(
    ///             TypeEq::new::<T>()    // `TypeEq<T, T>`
    ///                 .in_array(te_len) // `TypeEq<[T; LEN], [T; 0]>`
    ///                 .to_left([])      // Goes from `[T; 0]` to `[T; LEN]`
    ///         )
    ///     } else if let TypeCmp::Eq(te_len) = Usize.equals(Usize) {
    ///         // this branch is ran on `LEN == 1`
    ///         // `te_len` is inferred to be `TypeEq<Usize<LEN>, Usize<1>>`
    ///         Ok(TypeEq::NEW.in_array(te_len).to_left([pair.0]))
    ///     } else if let TypeCmp::Eq(te_len) = Usize.equals(Usize) {
    ///         // this branch is ran on `LEN == 2`
    ///         // `te_len` is inferred to be `TypeEq<Usize<LEN>, Usize<2>>`
    ///         Ok(TypeEq::NEW.in_array(te_len).to_left([pair.0, pair.1]))
    ///     } else {
    ///         Err(pair)
    ///     }
    /// }
    /// 
    /// ```
    /// 
    /// ### Struct
    /// 
    /// This example demonstrates how `Usize` can be used to pass a 
    /// const-generic struct to a function expecting a concrete type of that struct.
    /// 
    /// ```rust
    /// use typewit::{const_marker::Usize, TypeCmp};
    /// 
    /// assert_eq!(mutate(Array([])), Array([]));
    /// assert_eq!(mutate(Array([3])), Array([3]));
    /// assert_eq!(mutate(Array([3, 5])), Array([3, 5]));
    /// assert_eq!(mutate(Array([3, 5, 8])), Array([8, 5, 3])); // reversed!
    /// assert_eq!(mutate(Array([3, 5, 8, 13])), Array([3, 5, 8, 13]));
    /// 
    /// 
    /// #[derive(Debug, PartialEq)]
    /// struct Array<const CAP: usize>([u32; CAP]);
    /// 
    /// const fn mutate<const LEN: usize>(arr: Array<LEN>) -> Array<LEN> {
    ///     match Usize::<LEN>.equals(Usize::<3>) {
    ///         // `te_len` is a `TypeEq<Usize<LEN>, Usize<3>>`
    ///         // this branch is ran on `LEN == 3`
    ///         TypeCmp::Eq(te_len) => {
    ///             // `te` is a `TypeEq<Array<LEN>, Array<3>>`
    ///             let te = te_len.project::<GArray>();
    /// 
    ///             // `te.to_right(...)` here goes from `Array<LEN>` to `Array<3>`
    ///             let ret = reverse3(te.to_right(arr));
    /// 
    ///             // `te.to_left(...)` here goes from `Array<3>` to `Array<LEN>`
    ///             te.to_left(ret)
    ///         }
    ///         TypeCmp::Ne(_) => arr,
    ///     }
    /// }
    /// 
    /// const fn reverse3(Array([a, b, c]): Array<3>) -> Array<3> {
    ///     Array([c, b, a])
    /// }
    /// 
    /// typewit::type_fn!{
    ///     // Type-level function from `Usize<LEN>` to `Array<LEN>`
    ///     struct GArray;
    /// 
    ///     impl<const LEN: usize> Usize<LEN> => Array<LEN>
    /// }
    /// ```
    fn equals;
}

declare_const_param_type!{I8(i8)}
declare_const_param_type!{I16(i16)}
declare_const_param_type!{I32(i32)}
declare_const_param_type!{I64(i64)}
declare_const_param_type!{I128(i128)}
declare_const_param_type!{Isize(isize)}

