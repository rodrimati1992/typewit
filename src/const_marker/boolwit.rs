use core::fmt::{self, Debug};

use crate::{
    const_marker::Bool,
    TypeCmp,
    TypeEq,
    TypeWitnessTypeArg, MakeTypeWitness,
};


/// Type Witness that [`Bool<B>`](Bool) is either `Bool<true>` or `Bool<false>`.
/// 
/// Use this over [`BoolWitG`] if you have a `const B: bool` parameter already.
/// 
/// # Example
/// 
/// Making a function that takes a generic `Foo<B>` and calls methods on 
/// `Foo<false>` or `Foo<true>` depending on the value of the `const B: bool` parameter.
/// 
/// ```rust
/// use typewit::{const_marker::{Bool, BoolWit}, MakeTypeWitness};
/// 
/// 
/// assert_eq!(call_next(Incrementor::<GO_UP>(4)), Incrementor(5));
/// assert_eq!(call_next(Incrementor::<GO_UP>(5)), Incrementor(6));
/// 
/// assert_eq!(call_next(Incrementor::<GO_DOWN>(4)), Incrementor(3));
/// assert_eq!(call_next(Incrementor::<GO_DOWN>(3)), Incrementor(2));
/// 
/// 
/// const fn call_next<const B: bool>(incrementor: Incrementor<B>) -> Incrementor<B> {
///     typewit::type_fn! {
///         // type-level function from `Bool<B>` to `Incrementor<B>`
///         struct IncrementorFn;
///         impl<const B: bool> Bool<B> => Incrementor<B>
///     }
/// 
///     // The example below this one shows how to write this match more concisely
///     match BoolWit::MAKE {
///         // `bw: TypeEq<Bool<B>, Bool<true>>`
///         BoolWit::True(bw) => {
///             // `te: TypeEq<Incrementor<B>, Incrementor<true>>`
///             let te = bw.project::<IncrementorFn>();
/// 
///             // `te.to_right` casts `Incrementor<B>` to `Incrementor<true>`,
///             // (this allows calling the inherent method).
///             // 
///             // `te.to_left` casts `Incrementor<true>` to `Incrementor<B>`
///             te.to_left(te.to_right(incrementor).next())
///         }
///         // `bw: TypeEq<Bool<B>, Bool<false>>`
///         BoolWit::False(bw) => {
///             // `te: TypeEq<Incrementor<B>, Incrementor<false>>`
///             let te = bw.project::<IncrementorFn>();
/// 
///             // like the other branch, but with `Incrementor<false>`
///             te.to_left(te.to_right(incrementor).next())
///         }
///     }
/// }
/// 
/// 
/// #[derive(Debug, Copy, Clone, PartialEq, Eq)]
/// struct Incrementor<const GO_UP: bool>(usize);
/// 
/// const GO_UP: bool = true;
/// const GO_DOWN: bool = false;
/// 
/// impl Incrementor<GO_DOWN> {
///     #[track_caller]
///     pub const fn next(self) -> Self {
///         Self(self.0 - 1)
///     }
/// }
/// 
/// impl Incrementor<GO_UP> {
///     pub const fn next(self) -> Self {
///         Self(self.0 + 1)
///     }
/// }
/// 
/// ```
/// 
/// ### Using `polymatch` for conciseness
/// 
/// The [`polymatch`](crate::polymatch) macro can be used to 
/// more concisely implement the `call_next` function.
/// 
/// ```
/// # use typewit::{const_marker::{Bool, BoolWit}, MakeTypeWitness};
/// # 
/// const fn call_next<const B: bool>(incrementor: Incrementor<B>) -> Incrementor<B> {
///     typewit::type_fn! {
///         struct IncrementorFn;
///         impl<const B: bool> Bool<B> => Incrementor<B>
///     }
/// 
///     // expands to a match with two arms, 
///     // one for `BoolWit::True` and one for `BoolWit::False`,
///     // copying the expression to the right of the `=>` to both arms.
///     typewit::polymatch! {BoolWit::MAKE;
///         BoolWit::True(bw) | BoolWit::False(bw) => {
///             let te = bw.project::<IncrementorFn>();
///             te.to_left(te.to_right(incrementor).next())
///         }
///     }
/// }
/// # 
/// # #[derive(Debug, Copy, Clone, PartialEq, Eq)]
/// # struct Incrementor<const GO_UP: bool>(usize);
/// # 
/// # const GO_UP: bool = true;
/// # const GO_DOWN: bool = false;
/// # 
/// # impl Incrementor<GO_DOWN> {
/// #     #[track_caller]
/// #     pub const fn next(self) -> Self { unimplemented!() }
/// # }
/// # 
/// # impl Incrementor<GO_UP> {
/// #     pub const fn next(self) -> Self { unimplemented!() }
/// # }
/// ```
/// 
/// ### What happens without `BoolWit`
/// 
/// If the `call_next` function was defined like this:
/// ```rust,compile_fail
/// # use typewit::{const_marker::{Bool, BoolWit}, MakeTypeWitness};
/// # 
/// const fn call_next<const B: bool>(incrementor: Incrementor<B>) -> Incrementor<B> {
///     incrementor.next()
/// }
/// # #[derive(Copy, Clone)]
/// # struct Incrementor<const WRAPPING: bool>(usize);
/// # 
/// # impl Incrementor<false> {
/// #     pub const fn next(self) -> Self {
/// #         unimplemented!()
/// #     }
/// # }
/// # 
/// # impl Incrementor<true> {
/// #     pub const fn next(self) -> Self {
/// #         unimplemented!()
/// #     }
/// # }
/// ```
/// it would produce this error
/// ```text
/// error[E0599]: no method named `next` found for struct `Incrementor<B>` in the current scope
///   --> src/const_marker/const_witnesses.rs:20:17
///    |
/// 7  |     incrementor.next()
///    |                 ^^^^ method not found in `Incrementor<B>`
/// ...
/// 38 | struct Incrementor<const WRAPPING: bool>(usize);
///    | ---------------------------------------- method `next` not found for this struct
///    |
///    = note: the method was found for
///            - `Incrementor<false>`
///            - `Incrementor<true>`
/// ```
/// 
/// 
pub type BoolWit<const B: bool> = BoolWitG<Bool<B>>;


/// Type witness that `B` is either [`Bool`]`<true>` or [`Bool`]`<false>`
/// 
/// Use this over [`BoolWit`] if you want to write a [`HasTypeWitness`] bound
/// and adding a `const B: bool` parameter would be impossible.
/// 
/// # Example
/// 
/// This basic example demonstrates where `BoolWitG` would be used instead of `BoolWit`.
/// 
/// ```rust
/// use typewit::const_marker::{Bool, BoolWitG};
/// use typewit::HasTypeWitness;
/// 
/// 
/// trait Boolean: Sized + HasTypeWitness<BoolWitG<Self>> {
///     type Not: Boolean<Not = Self>;
/// }
/// 
/// impl Boolean for Bool<true> {
///     type Not = Bool<false>;
/// }
/// 
/// impl Boolean for Bool<false> {
///     type Not = Bool<true>;
/// }
/// ```
/// 
/// [`HasTypeWitness`]: crate::HasTypeWitness
pub enum BoolWitG<B> {
    /// Witnesses that `B == true`
    True(TypeEq<B, Bool<true>>),
    /// Witnesses that `B == false`
    False(TypeEq<B, Bool<false>>),
}

impl<B> BoolWitG<B> {
    /// Whether `B == Bool<true>`
    ///
    /// # Example
    ///
    /// ```rust
    /// use typewit::{const_marker::BoolWitG, TypeEq};
    /// 
    /// assert_eq!(BoolWitG::True(TypeEq::NEW).is_true(), true);
    /// assert_eq!(BoolWitG::False(TypeEq::NEW).is_true(), false);
    /// ```
    ///
    pub const fn is_true(self) -> bool {
        matches!(self, Self::True{..})
    }

    /// Whether `B == Bool<false>`
    ///
    /// # Example
    ///
    /// ```rust
    /// use typewit::{const_marker::BoolWitG, TypeEq};
    /// 
    /// assert_eq!(BoolWitG::True(TypeEq::NEW).is_false(), false);
    /// assert_eq!(BoolWitG::False(TypeEq::NEW).is_false(), true);
    /// ```
    ///
    pub const fn is_false(self) -> bool {
        matches!(self, Self::False{..})
    }

    /// Gets a proof of `B == Bool<true>`, returns None if `B == Bool<false>`
    ///
    /// # Example
    ///
    /// ```rust
    /// use typewit::{const_marker::{Bool, BoolWitG}, TypeEq};
    /// 
    /// assert_eq!(BoolWitG::True(TypeEq::NEW).to_true(), Some(TypeEq::new::<Bool<true>>()));
    /// assert_eq!(BoolWitG::False(TypeEq::NEW).to_true(), None);
    /// ```
    ///
    pub const fn to_true(self) -> Option<TypeEq<B, Bool<true>>> {
        match self {
            Self::True(x) => Some(x),
            Self::False{..} => None
        }
    }

    /// Gets a proof of `B == Bool<false>`, returns None if `B == Bool<true>`
    ///
    /// # Example
    ///
    /// ```rust
    /// use typewit::{const_marker::{Bool, BoolWitG}, TypeEq};
    /// 
    /// assert_eq!(BoolWitG::True(TypeEq::NEW).to_false(), None);
    /// assert_eq!(BoolWitG::False(TypeEq::NEW).to_false(), Some(TypeEq::new::<Bool<false>>()));
    /// ```
    ///
    pub const fn to_false(self) -> Option<TypeEq<B, Bool<false>>> {
        match self {
            Self::False(x) => Some(x),
            Self::True{..} => None
        }
    }


    /// Gets a proof of `B == Bool<true>`.
    ///
    /// # Panic
    ///
    /// Panics if `B == Bool<false>`
    ///
    /// # Example
    ///
    /// ```rust
    /// use typewit::{const_marker::{Bool, BoolWitG}, TypeEq};
    /// 
    /// assert_eq!(BoolWitG::True(TypeEq::NEW).unwrap_true(), TypeEq::new::<Bool<true>>());
    /// ```
    ///
    pub const fn unwrap_true(self) -> TypeEq<B, Bool<true>> {
        match self {
            Self::True(x) => x,
            Self::False{..}  => panic!("attempted to unwrap into True on False variant")
        }
    }

    /// Gets a proof of `B == Bool<false>`.
    ///
    /// # Panic
    /// 
    /// Panics if `B == Bool<true>`
    ///
    /// # Example
    ///
    /// ```rust
    /// use typewit::{const_marker::{Bool, BoolWitG}, TypeEq};
    /// 
    /// assert_eq!(BoolWitG::False(TypeEq::NEW).unwrap_false(), TypeEq::new::<Bool<false>>());
    /// ```
    ///
    pub const fn unwrap_false(self) -> TypeEq<B, Bool<false>> {
        match self {
            Self::False(x) => x,
            Self::True{..} => panic!("attempted to unwrap into False on True variant")
        }
    }

}

impl<B> Copy for BoolWitG<B> {}

impl<B> Clone for BoolWitG<B> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<B> Debug for BoolWitG<B> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str(match self {
            Self::True{..} => "True",
            Self::False{..} => "False",
        })
    }
}

impl<B> TypeWitnessTypeArg for BoolWitG<B> {
    type Arg = B;
}

impl<const B: bool> MakeTypeWitness for BoolWitG<Bool<B>> {
    const MAKE: Self = {
        if let TypeCmp::Eq(te) = Bool.equals(Bool) {
            BoolWit::True(te)
        } else if let TypeCmp::Eq(te) = Bool.equals(Bool) {
            BoolWit::False(te)
        } else {
            panic!("unreachable: `B` is either `true` or `false`")
        }
    };
}

