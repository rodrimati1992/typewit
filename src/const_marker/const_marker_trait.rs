/// A type that represents a constant
/// 
/// # Example
/// 
/// Emulating generic associated constants on stable
/// 
/// ```rust
/// use typewit::const_marker::{ConstMarker, ConstMarkerOf};
/// 
/// 
/// assert_eq!(make_array::<u8>(), ([3], [3, 3, 3]));
/// 
/// assert_eq!(make_array::<&str>(), (["hi"], ["hi", "hi", "hi"]));
/// 
/// const fn make_array<T: MakeArray>() -> ([T; 1], [T; 3]) {
///     (T::Make::<1>::VAL, T::Make::<3>::VAL)
/// }
/// 
/// // `MakeArray` can make arrays of any length without writing a bound for each length!!
/// trait MakeArray: Sized {
///     // emulates `const Make<const N: usize>: [Self; N];` (a generic associated constant)
///     type Make<const N: usize>: ConstMarkerOf<[Self; N]>;
/// }
/// 
/// 
/// impl MakeArray for u8 {
///     type Make<const N: usize> = MakeArrayConst<Self, N>;
/// }
///
/// impl<'a> MakeArray for &'a str {
///     type Make<const N: usize> = MakeArrayConst<Self, N>;
/// }
/// 
/// struct MakeArrayConst<T, const N: usize>(core::marker::PhantomData<T>);
/// 
/// impl<const N: usize> ConstMarker for MakeArrayConst<u8, N> {
///     type Of = [u8; N];
///     const VAL: Self::Of = [3; N];
/// }
/// 
/// impl<'a, const N: usize> ConstMarker for MakeArrayConst<&'a str, N> {
///     type Of = [&'a str; N];
///     const VAL: Self::Of = ["hi"; N];
/// }
/// 
/// ```
/// 
pub trait ConstMarker: Sized {
    /// The type of this constant
    type Of;

    /// The value of this constant
    const VAL: Self::Of;
}


/// Trait alias for `ConstMarker<Of = Of>`
/// 
/// This trait also shows the `Of` type in unsatisfied trait bound errors
/// (`ConstMarker<Of = ...>` bounds obfuscate this a bit as of 1.89.0) 
#[cfg_attr(feature = "rust_1_83", diagnostic::on_unimplemented(
    message = "`{Self}` is not a type-level constant of type `{Of}`",
))]
pub trait ConstMarkerOf<Of>: ConstMarker<Of = Of> {}

impl<C, Of> ConstMarkerOf<Of> for C
where
    C: ConstMarker<Of = Of>,
{}
