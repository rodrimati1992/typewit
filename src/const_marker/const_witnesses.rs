use crate::{
    const_marker::Bool,
    TypeCmp,
    TypeEq,
    TypeWitnessTypeArg, MakeTypeWitness,
};



/// Type Witness that [`Bool<B>`](Bool) is either `Bool<true>` or `Bool<false>`.
/// 
/// # Example
/// 
/// Making a function that takes a generic `Foo<B>` and calls methods on 
/// `Foo<false>` or `Foo<true>` depending on the value of the `const B: bool` parameter.
/// 
/// ```rust
/// use typewit::{const_marker::{Bool, BoolWit}, MakeTypeWitness};
/// 
/// 
/// assert_eq!(call_next(Incrementor::<GO_UP>(4)), Incrementor(5));
/// assert_eq!(call_next(Incrementor::<GO_UP>(5)), Incrementor(6));
/// 
/// assert_eq!(call_next(Incrementor::<GO_DOWN>(4)), Incrementor(3));
/// assert_eq!(call_next(Incrementor::<GO_DOWN>(3)), Incrementor(2));
/// 
/// 
/// const fn call_next<const B: bool>(incrementor: Incrementor<B>) -> Incrementor<B> {
///     typewit::type_fn! {
///         // type-level function from `Bool<B>` to `Incrementor<B>`
///         struct IncrementorFn;
///         impl<const B: bool> Bool<B> => Incrementor<B>
///     }
/// 
///     // The example below this one shows how to write this match more concisely
///     match BoolWit::MAKE {
///         // `bw: TypeEq<Bool<B>, Bool<true>>`
///         BoolWit::True(bw) => {
///             // `te: TypeEq<Incrementor<B>, Incrementor<true>>`
///             let te = bw.project::<IncrementorFn>();
/// 
///             // `te.to_right` casts `Incrementor<B>` to `Incrementor<true>`,
///             // (this allows calling the inherent method).
///             // 
///             // `te.to_left` casts `Incrementor<true>` to `Incrementor<B>`
///             te.to_left(te.to_right(incrementor).next())
///         }
///         // `bw: TypeEq<Bool<B>, Bool<false>>`
///         BoolWit::False(bw) => {
///             // `te: TypeEq<Incrementor<B>, Incrementor<false>>`
///             let te = bw.project::<IncrementorFn>();
/// 
///             // like the other branch, but with `Incrementor<false>`
///             te.to_left(te.to_right(incrementor).next())
///         }
///     }
/// }
/// 
/// 
/// #[derive(Debug, Copy, Clone, PartialEq, Eq)]
/// struct Incrementor<const GO_UP: bool>(usize);
/// 
/// const GO_UP: bool = true;
/// const GO_DOWN: bool = false;
/// 
/// impl Incrementor<GO_DOWN> {
///     #[track_caller]
///     pub const fn next(self) -> Self {
///         Self(self.0 - 1)
///     }
/// }
/// 
/// impl Incrementor<GO_UP> {
///     pub const fn next(self) -> Self {
///         Self(self.0 + 1)
///     }
/// }
/// 
/// ```
/// 
/// ### Using `polymatch` for conciseness
/// 
/// The [`polymatch`](crate::polymatch) macro can be used to 
/// more concisely implement the `call_next` function.
/// 
/// ```
/// # use typewit::{const_marker::{Bool, BoolWit}, MakeTypeWitness};
/// # 
/// const fn call_next<const B: bool>(incrementor: Incrementor<B>) -> Incrementor<B> {
///     typewit::type_fn! {
///         struct IncrementorFn;
///         impl<const B: bool> Bool<B> => Incrementor<B>
///     }
/// 
///     // expands to a match with two arms, 
///     // one for `BoolWit::True` and one for `BoolWit::False`,
///     // copying the expression to the right of the `=>` to both arms.
///     typewit::polymatch! {BoolWit::MAKE;
///         BoolWit::True(bw) | BoolWit::False(bw) => {
///             let te = bw.project::<IncrementorFn>();
///             te.to_left(te.to_right(incrementor).next())
///         }
///     }
/// }
/// # 
/// # #[derive(Debug, Copy, Clone, PartialEq, Eq)]
/// # struct Incrementor<const GO_UP: bool>(usize);
/// # 
/// # const GO_UP: bool = true;
/// # const GO_DOWN: bool = false;
/// # 
/// # impl Incrementor<GO_DOWN> {
/// #     #[track_caller]
/// #     pub const fn next(self) -> Self { unimplemented!() }
/// # }
/// # 
/// # impl Incrementor<GO_UP> {
/// #     pub const fn next(self) -> Self { unimplemented!() }
/// # }
/// ```
/// 
/// ### What happens without `BoolWit`
/// 
/// If the `call_next` function was defined like this:
/// ```rust,compile_fail
/// # use typewit::{const_marker::{Bool, BoolWit}, MakeTypeWitness};
/// # 
/// const fn call_next<const B: bool>(incrementor: Incrementor<B>) -> Incrementor<B> {
///     incrementor.next()
/// }
/// # #[derive(Copy, Clone)]
/// # struct Incrementor<const WRAPPING: bool>(usize);
/// # 
/// # impl Incrementor<false> {
/// #     pub const fn next(self) -> Self {
/// #         unimplemented!()
/// #     }
/// # }
/// # 
/// # impl Incrementor<true> {
/// #     pub const fn next(self) -> Self {
/// #         unimplemented!()
/// #     }
/// # }
/// ```
/// it would produce this error
/// ```text
/// error[E0599]: no method named `next` found for struct `Incrementor<B>` in the current scope
///   --> src/const_marker/const_witnesses.rs:20:17
///    |
/// 7  |     incrementor.next()
///    |                 ^^^^ method not found in `Incrementor<B>`
/// ...
/// 38 | struct Incrementor<const WRAPPING: bool>(usize);
///    | ---------------------------------------- method `next` not found for this struct
///    |
///    = note: the method was found for
///            - `Incrementor<false>`
///            - `Incrementor<true>`
/// ```
/// 
/// 
pub enum BoolWit<const B: bool> {
    /// Witnesses that `B == true`
    True(TypeEq<Bool<B>, Bool<true>>),
    /// Witnesses that `B == false`
    False(TypeEq<Bool<B>, Bool<false>>),
}

impl<const B: bool> TypeWitnessTypeArg for BoolWit<B> {
    type Arg = Bool<B>;
}

impl<const B: bool> MakeTypeWitness for BoolWit<B> {
    const MAKE: Self = {
        if let TypeCmp::Eq(te) = Bool.equals(Bool) {
            BoolWit::True(te)
        } else if let TypeCmp::Eq(te) = Bool.equals(Bool) {
            BoolWit::False(te)
        } else {
            panic!("unreachable: `B` is either `true` or `false`")
        }
    };
}

