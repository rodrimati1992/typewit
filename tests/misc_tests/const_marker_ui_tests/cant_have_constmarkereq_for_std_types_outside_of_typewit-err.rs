use typewit::TypeCmp;
use typewit::const_marker::{ConstMarker, ConstMarkerEq};


struct Foo<const N: u8>;

impl<const N: u8> ConstMarker for Foo<N> {
    type Of = u8;
    const VAL: u8 = N;
}

impl<const N: u8> ConstMarkerEq for Foo<N> {
    type Equals<Rhs: ConstMarkerEq<Of = Self::Of>> = DumbEquals<Self, Rhs>;
}

struct DumbEquals<L, R>(L, R);

impl<L, R> ConstMarker for DumbEquals<L, R> {
    type Of = TypeCmp<L, R>;
    const VAL: Self::Of = unimplemented!();
}


fn main() {}
