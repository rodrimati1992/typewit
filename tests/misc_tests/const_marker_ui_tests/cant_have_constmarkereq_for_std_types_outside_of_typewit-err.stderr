error[E0277]: the trait bound `__StdIdentityTypeEq<Foo<N>>: MakeTypeWitness` is not satisfied
  --> tests/misc_tests/const_marker_ui_tests/cant_have_constmarkereq_for_std_types_outside_of_typewit-err.rs:12:37
   |
12 | impl<const N: u8> ConstMarkerEq for Foo<N> {
   |                                     ^^^^^^ the trait `MakeTypeWitness` is not implemented for `__StdIdentityTypeEq<Foo<N>>`
   |
   = help: the following other types implement trait `MakeTypeWitness`:
             __StdIdentityTypeEq<Bool<T>>
             __StdIdentityTypeEq<I128<T>>
             __StdIdentityTypeEq<I16<T>>
             __StdIdentityTypeEq<I32<T>>
             __StdIdentityTypeEq<I64<T>>
             __StdIdentityTypeEq<I8<T>>
             __StdIdentityTypeEq<Isize<T>>
             __StdIdentityTypeEq<U128<T>>
           and $N others
   = note: required for `Foo<N>` to implement `ConstMarkerHasWitness`
note: required by a bound in `ConstMarkerEq`
  --> src/const_marker/const_marker_eq_traits.rs
   |
   | pub trait ConstMarkerEq: ConstMarker + ConstMarkerHasWitness {
   |                                        ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `ConstMarkerEq`
